File: src/bot/interaction-handler.ts
import { Interaction, ChatInputCommandInteraction, ButtonInteraction, MessageFlags } from 'discord.js';
import { GuildRepository, ConfirmationRepository, ExecutionRepository, AuditRepository } from '../database/repositories.js';
import { ActionDispatcher } from '../actions/dispatcher.js';
import { DiscordUIComponents } from './ui.js';
import { createLLMProvider } from '../ai/factory.js';
import { FallbackParser } from '../ai/fallback-parser.js';
import { IntentStatus } from '../ai/types.js';
import { InjectionDefense } from '../security/injection-defense.js';
import { EntityResolver } from '../entities/entity-resolver.js';
import { logger } from '../utils/logger.js';

export class InteractionHandler {
  private llmProvider = createLLMProvider();

  async handleInteraction(interaction: Interaction) {
    try {
      if (interaction.isChatInputCommand()) {
        await this.handleChatInputCommand(interaction);
      } else if (interaction.isButton()) {
        await this.handleButtonInteraction(interaction);
      }
    } catch (err: any) {
      // 10062 = interaction token expired (>3s without ack, or user deleted message).
      // Retrying a reply only produces a second 10062, so just log and stop.
      if (err?.code === 10062 || err?.rawError?.code === 10062) {
        logger.warn('Interaction expired before bot could respond (Unknown interaction). Likely slow ack.');
        return;
      }
      logger.error({ err }, 'Error handling interaction');
      try {
        if (interaction.isRepliable()) {
          const errorUI = DiscordUIComponents.createErrorEmbed('System Error', err.message || 'An unexpected error occurred.');
          if (interaction.deferred || interaction.replied) {
            await interaction.followUp({ ...errorUI, flags: MessageFlags.Ephemeral });
          } else {
            await interaction.reply({ ...errorUI, flags: MessageFlags.Ephemeral });
          }
        }
      } catch (replyErr: any) {
        if (replyErr?.code === 10062) {
          logger.warn('Could not send error message: interaction already expired.');
        } else {
          logger.error({ err: replyErr }, 'Failed to send error response');
        }
      }
    }
  }

  private async handleChatInputCommand(interaction: ChatInputCommandInteraction) {
    if (!interaction.guild || !interaction.member) {
      await interaction.reply({ content: 'Commands can only be used within a server.', flags: MessageFlags.Ephemeral });
      return;
    }

    // ACK FIRST: Discord invalidates the interaction token after 3s.
    // All slow work (member fetch, DB, LLM) must happen after defer.
    const needsEphemeral = interaction.commandName === 'prompt-audit' || interaction.commandName === 'prompt-help';
    try {
      await interaction.deferReply(needsEphemeral ? { flags: MessageFlags.Ephemeral } : undefined);
    } catch (err: any) {
      if (err?.code === 10062) {
        logger.warn('Interaction already expired before defer. Aborting.');
        return;
      }
      throw err;
    }

    const guildId = interaction.guildId!;
    const userId = interaction.user.id;
    const guild = interaction.guild;
    const startedAt = Date.now();
    const ctxBase = { command: interaction.commandName, guildId, userId };
    const stage = (name: string) =>
      logger.info({ ...ctxBase, elapsedMs: Date.now() - startedAt }, `prompt stage: ${name}`);
    stage('deferred - fetching member');

    const member = await guild.members.fetch(userId);
    stage('member fetched - loading guild settings');

    const guildDb = await GuildRepository.findOrCreate(guildId, guild.name);
    const settings = guildDb.settings;
    stage('guild ready');

    if (interaction.commandName === 'prompt') {
      const rawPrompt = interaction.options.getString('request', true);
      const prompt = InjectionDefense.sanitizeExternalText(rawPrompt);
      stage(`prompt received (${prompt.length} chars) - injection check`);

      if (InjectionDefense.detectPromptInjection(prompt)) {
        const errUI = DiscordUIComponents.createErrorEmbed(
          'Security Violation',
          'Prompt contains suspicious system/injection instructions. Request rejected.'
        );
        await interaction.editReply(errUI);
        return;
      }

      let parsed = FallbackParser.parse(prompt);

      if (!parsed) {
        stage('fallback miss - fetching channels/roles for LLM context');
        const channels = (await guild.channels.fetch()).map((c) => ({ id: c?.id || '', name: c?.name || '' }));
        const roles = (await guild.roles.fetch()).map((r) => ({ id: r.id, name: r.name }));

        stage(`context ready (${channels.length} channels, ${roles.length} roles) - calling LLM`);
        // Keep Discord updated so it doesn't look stuck on "thinking".
        await interaction.editReply('Parsing your request with AI - this can take a few seconds…');
        parsed = await this.llmProvider.parseIntent(prompt, {
          guildId,
          channels,
          roles,
          allowedActions: settings?.allowedActions?.length ? settings.allowedActions : ['create_channel', 'delete_channel', 'create_role', 'assign_role', 'timeout_member', 'ban_member', 'send_message'],
        });
        stage(`LLM done - status=${parsed.status}`);
        logger.debug({ ...ctxBase, parsed: JSON.stringify(parsed).slice(0, 1000) }, 'parsed intent');
      } else {
        stage(`fallback hit - action=${parsed.action}`);
      }

      if (parsed.status === IntentStatus.UNSUPPORTED || parsed.status === IntentStatus.REJECTED) {
        stage(`rejected - replying (${parsed.reason?.slice(0, 80) ?? 'no reason'})`);
        const errUI = DiscordUIComponents.createErrorEmbed('Operation Rejected', parsed.reason || 'This request is unsupported or violates safety guidelines.');
        await interaction.editReply(errUI);
        return;
      }

      if (parsed.status === IntentStatus.CLARIFICATION_REQUIRED) {
        stage('clarification required - replying');
        const errUI = DiscordUIComponents.createErrorEmbed('Clarification Required', parsed.question || 'Please provide more details.');
        await interaction.editReply(errUI);
        return;
      }

      if (parsed.status === IntentStatus.DIRECT_ACTION && parsed.action) {
        stage(`dispatching action=${parsed.action}`);
        const execution = await ExecutionRepository.createExecution({
          guildId,
          userId,
          prompt,
          status: 'RUNNING',
          parsedIntent: parsed,
        });

        const ctx = {
          guildId,
          userId,
          executionId: execution.id,
          guild,
          actorMember: member,
          settings,
        };

        let actionType = parsed.action;
        let params = parsed.parameters || {};

        if (params.channelName) {
          const res = await EntityResolver.resolveChannel(guild, params.channelName);
          if (res.resolved) params.channelId = res.resolved.id;
        }

        const dispatchRes = await ActionDispatcher.dispatch(actionType, params, ctx);
        stage(
          dispatchRes.requiresConfirmation
            ? 'dispatch needs confirmation - replying'
            : dispatchRes.success
              ? `dispatch ok - replying`
              : `dispatch failed (${dispatchRes.error?.slice(0, 80) ?? 'unknown'}) - replying`
        );

        if (dispatchRes.requiresConfirmation && dispatchRes.confirmationDetails) {
          const ui = DiscordUIComponents.createConfirmationEmbed(
            dispatchRes.confirmationDetails.actionType,
            dispatchRes.confirmationDetails.riskLevel,
            JSON.stringify(params, null, 2),
            dispatchRes.confirmationDetails.actionNonce
          );
          await interaction.editReply(ui);
          return;
        }

        if (dispatchRes.success) {
          const ui = DiscordUIComponents.createSuccessEmbed('Action Executed', `Successfully performed **${actionType}**.`);
          await interaction.editReply(ui);
        } else {
          const ui = DiscordUIComponents.createErrorEmbed('Action Failed', dispatchRes.error || 'Execution failed.');
          await interaction.editReply(ui);
        }
      } else {
        // Catch-all: UNSUPPORTED/REJECTED/CLARIFICATION/DIRECT_ACTION are handled
        // above, but the model can also return action_plan, confirmation_required,
        // capability_unavailable, permission_analysis_required, or a direct_action
        // with no action. Without this the bot sits on "thinking" forever.
        stage(`unhandled intent (status=${parsed.status}, action=${parsed.action ?? 'none'}) - replying`);
        const detail =
          parsed.status === IntentStatus.DIRECT_ACTION
            ? 'The AI response did not include an executable action. Try rephrasing with a concrete operation.'
            : `The AI returned status "${parsed.status}", which this bot cannot execute yet. Try a single concrete operation (e.g. "Create channel welcome").`;
        await interaction.editReply(
          DiscordUIComponents.createErrorEmbed('Operation Rejected', parsed.reason || detail)
        );
      }
    } else if (interaction.commandName === 'prompt-audit') {
      const logs = await AuditRepository.getRecentLogs(guildId, 10);
      const logSummary = logs.map((l: { createdAt: Date; action: string; userId: string; status: string }) => `• [${l.createdAt.toISOString()}] **${l.action}** by <@${l.userId}>: ${l.status}`).join('\n') || 'No audit logs found.';
      await interaction.editReply({ embeds: [DiscordUIComponents.createSuccessEmbed('Recent Audit Logs', logSummary).embeds[0]] });
    } else if (interaction.commandName === 'prompt-help') {
      await interaction.editReply({
        embeds: [
          DiscordUIComponents.createSuccessEmbed(
            'Natural Language Bot Help',
            'You can describe operations using `/prompt`!\n\n**Supported Actions:**\n• Create Channel: `Create channel welcome`\n• Delete Channel: `Delete #old-chat`\n• Create Role: `Create role Moderators`\n• Assign Role: `Give @Alex role Support`\n• Timeout Member: `Timeout @BadActor for 300 seconds`\n• Ban Member: `Ban @Spammer`'
          ).embeds[0]
        ],
      });
    }
  }

  private async handleButtonInteraction(interaction: ButtonInteraction) {
    const customId = interaction.customId;
    if (customId.startsWith('confirm:') || customId.startsWith('cancel:')) {
      const [action, nonce] = customId.split(':');
      const guildId = interaction.guildId!;
      const userId = interaction.user.id;

      if (action === 'cancel') {
        await interaction.update(DiscordUIComponents.createErrorEmbed('Action Cancelled', 'The confirmation was explicitly cancelled by user.'));
        return;
      }

      // ACK before slow DB + Discord work (3s interaction window).
      // After deferUpdate, respond with editReply (update would throw).
      try {
        await interaction.deferUpdate();
      } catch (err: any) {
        if (err?.code === 10062) {
          logger.warn('Button interaction already expired before ack. Aborting.');
          return;
        }
        throw err;
      }

      const confirmation = await ConfirmationRepository.findAndConsume(nonce, guildId, userId, '');
      if (!confirmation) {
        await interaction.editReply(DiscordUIComponents.createErrorEmbed('Invalid or Expired Confirmation', 'This confirmation has expired, was already consumed, or belonged to another user.'));
        return;
      }

      const plan = confirmation.actionPlan as any;
      const guild = interaction.guild!;
      const member = await guild.members.fetch(userId);

      const ctx = {
        guildId,
        userId,
        executionId: `conf-${Date.now()}`,
        guild,
        actorMember: member,
      };

      const actionDef = (await import('../actions/registry.js')).ActionRegistry.get(plan.type);
      if (!actionDef) {
        await interaction.editReply(DiscordUIComponents.createErrorEmbed('Execution Failed', 'Action definition no longer exists.'));
        return;
      }

      try {
        await actionDef.handler(ctx, plan.input);
        await interaction.editReply(DiscordUIComponents.createSuccessEmbed('Action Executed', `Successfully executed confirmed operation **${plan.type}**.`));
      } catch (err: any) {
        await interaction.editReply(DiscordUIComponents.createErrorEmbed('Execution Failed', err.message));
      }
    }
  }
}

File: src/bot/interaction_handler.py
from __future__ import annotations

import logging
from datetime import UTC, datetime

import discord
from discord.ext import commands

from src.actions.dispatcher import ActionDispatcher
from src.actions.plan_executor import MAX_PLAN_STEPS, PLAN_TYPE, execute_action_plan, summarize_plan
from src.actions.registry import ActionRegistry
from src.actions.types import ExecutionContext
from src.ai.fallback_parser import FallbackParser
from src.bot.ui import (
    ClarifyModal,
    ConfirmationView,
    DisambiguationView,
    DiscordUIComponents,
)
from src.database.repositories import (
    ConfirmationRepository,
    ConversationRepository,
    ExecutionRepository,
    GuildRepository,
    PendingChoiceRepository,
    PendingQuestionRepository,
)
from src.database.session import AsyncSessionLocal
from src.security.injection_defense import InjectionDefense

logger = logging.getLogger("discord_agent")

CHANNEL_ACTIONS = {
    "send_message",
    "purge_messages",
    "set_slowmode",
    "lock_channel",
    "unlock_channel",
    "rename_channel",
    "set_topic",
    "pin_message",
    "unpin_message",
    "delete_channel",
    "edit_channel",
}


class BotInteractionHandler:
    def __init__(self, bot: commands.Bot):
        self.bot = bot

    # ---------- entry ----------

    async def handle_prompt(self, interaction: discord.Interaction, request: str):
        await interaction.response.defer(thinking=True, ephemeral=False)

        guild = interaction.guild
        if guild is None:
            await interaction.followup.send(
                "Commands can only be executed within a server.", ephemeral=True
            )
            return

        guild_id = str(guild.id)
        user_id = str(interaction.user.id)
        prompt = InjectionDefense.sanitize_external_text(request)

        if InjectionDefense.detect_prompt_injection(prompt):
            await interaction.followup.send(
                embed=DiscordUIComponents.create_error_embed(
                    "Security Violation",
                    "Prompt contains suspicious system/injection instructions. Request rejected.",
                )
            )
            return

        async with AsyncSessionLocal() as session:
            # Upsert side effect only: the canonical tenant key is the
            # Discord snowflake (guild_id), matching the legacy FK targets.
            await GuildRepository.find_or_create(session, guild_id, guild.name)
            settings = await GuildRepository.get_settings(session, guild_id)
            channel_id = str(getattr(interaction, "channel_id", None) or guild_id)
            convo = await ConversationRepository.get_or_create(
                session, guild_id, user_id, channel_id
            )
            history = await ConversationRepository.history_lines(session, convo.id)

            parsed = FallbackParser.parse(prompt)
            source = "fallback"
            if parsed is None:
                from src.ai.needle_agent import parse_with_needle

                now = datetime.now(UTC).strftime("%Y-%m-%d %a %H:%M")
                channels = [{"id": str(c.id), "name": c.name} for c in guild.channels[:100]]
                roles = [{"id": str(r.id), "name": r.name} for r in guild.roles[:100]]
                allowed = None
                if settings is not None:
                    allow = getattr(settings, "allowedActions", None) or []
                    if allow:
                        allowed = list(allow)
                system = (
                    f"date: {now}; locale: en-US; channels: {len(channels)}; roles: {len(roles)}"
                )
                if history:
                    system += "; history: " + " | ".join(history[-6:])
                parsed = parse_with_needle(prompt, allowed_actions=allowed, system=system)
                source = "needle"
                logger.info("prompt parsed via=%s status=%s", source, parsed.get("status"))

            status = parsed.get(
                "status", "direct_action" if parsed.get("action") else "unsupported"
            )

            if status in ("unsupported", "rejected"):
                reason = parsed.get("reason", "This request is unsupported.")
                await interaction.followup.send(
                    embed=DiscordUIComponents.create_error_embed("Operation Rejected", reason)
                )
                await ConversationRepository.append(session, convo.id, "user", prompt)
                await ConversationRepository.append(
                    session, convo.id, "assistant", f"I couldn't do that: {reason[:300]}"
                )
                return

            if status == "chat":
                text = str(parsed.get("message", "")).strip()[:1900] or "How can I help?"
                await interaction.followup.send(text)
                await ConversationRepository.append(session, convo.id, "user", prompt)
                await ConversationRepository.append(session, convo.id, "assistant", text[:500])
                return

            if status == "clarification_required":
                question = str(parsed.get("question") or "Please provide more details.")[:500]
                pending = await PendingQuestionRepository.create(
                    session,
                    guild_id=guild_id,
                    user_id=user_id,
                    question=question,
                    original_prompt=prompt,
                    channel_id=channel_id,
                )
                embed = DiscordUIComponents.create_question_embed(question)
                view = discord.ui.View(timeout=900)
                answer_btn = discord.ui.Button(label="Answer", style=discord.ButtonStyle.primary)

                async def _ask(itx: discord.Interaction):
                    await itx.response.send_modal(
                        ClarifyModal(
                            title="Additional detail",
                            on_submit_cb=lambda m_itx, ans: self._resume_after_answer(
                                m_itx, pending.id, ans
                            ),
                        )
                    )

                answer_btn.callback = _ask  # type: ignore[method-assign]
                view.add_item(answer_btn)
                await interaction.followup.send(embed=embed, view=view)
                await ConversationRepository.append(session, convo.id, "user", prompt)
                await ConversationRepository.append(
                    session, convo.id, "assistant", f"I asked: {question[:300]}"
                )
                return

            if status == "action_plan":
                steps = parsed.get("steps") or []
                if len(steps) > MAX_PLAN_STEPS:
                    msg = (
                        f"That plan has {len(steps)} steps; I execute at most {MAX_PLAN_STEPS}. "
                        "Break it into smaller requests."
                    )
                    await interaction.followup.send(
                        embed=DiscordUIComponents.create_error_embed("Plan Too Large", msg)
                    )
                    return
                await self._run_plan(
                    interaction,
                    session,
                    guild,
                    guild_id,
                    user_id,
                    prompt,
                    convo.id,
                    steps,
                    settings,
                )
                return

            # direct_action (fallback shape {"action", "parameters"} or needle shape)
            action_type = parsed.get("action", "")
            params = dict(parsed.get("parameters") or {})
            if not action_type:
                await interaction.followup.send(
                    embed=DiscordUIComponents.create_error_embed(
                        "Operation Rejected", "No executable action found. Try rephrasing."
                    )
                )
                return
            await self._execute_single_action(
                interaction,
                session,
                guild,
                guild_id,
                user_id,
                prompt,
                convo.id,
                action_type,
                params,
                settings,
                channel_id,
            )

    # ---------- single action ----------

    async def _execute_single_action(
        self,
        interaction,
        session,
        guild,
        tenant_id,
        user_id,
        prompt,
        convo_id,
        action_type,
        params,
        settings,
        default_channel_id,
    ):
        from src.entities.entity_resolver import EntityResolver

        # Channel name -> id
        if params.get("channel_name") and not params.get("channel_id"):
            res = await EntityResolver.resolve_channel(guild, str(params["channel_name"]))
            if res.resolved is not None:
                params["channel_id"] = str(res.resolved.id)
            elif res.ambiguous and res.matches:
                await self._ask_disambiguation(
                    interaction,
                    session,
                    guild,
                    tenant_id,
                    user_id,
                    prompt,
                    convo_id,
                    action_type,
                    params,
                    "channel_id",
                    [{"id": str(c.id), "name": f"#{c.name}"} for c in res.matches[:10]],
                    "channel",
                    default_channel_id,
                    settings,
                )
                return
            else:
                msg = f'I couldn\'t find a channel named "{params["channel_name"]}". Check the name and try again.'
                await self._reply_error(
                    interaction, session, convo_id, prompt, "Channel Not Found", msg
                )
                return

        if params.get("role_name") and not params.get("role_id"):
            res = await EntityResolver.resolve_role(guild, str(params["role_name"]))
            if res.resolved is not None:
                params["role_id"] = str(res.resolved.id)
            elif res.ambiguous and res.matches:
                await self._ask_disambiguation(
                    interaction,
                    session,
                    guild,
                    tenant_id,
                    user_id,
                    prompt,
                    convo_id,
                    action_type,
                    params,
                    "role_id",
                    [{"id": str(r.id), "name": f"@{r.name}"} for r in res.matches[:10]],
                    "role",
                    default_channel_id,
                    settings,
                )
                return
            else:
                msg = f'I couldn\'t find a role named "{params["role_name"]}". Check the name and try again.'
                await self._reply_error(
                    interaction, session, convo_id, prompt, "Role Not Found", msg
                )
                return

        if not params.get("channel_id") and action_type in CHANNEL_ACTIONS and default_channel_id:
            params["channel_id"] = default_channel_id

        execution = await ExecutionRepository.create_execution(
            session=session,
            guild_id=tenant_id,
            user_id=user_id,
            prompt=prompt,
            status="RUNNING",
            parsed_intent={"action": action_type, "parameters": params},
        )
        actor = interaction.user
        if not isinstance(actor, discord.Member):
            try:
                actor = await guild.fetch_member(interaction.user.id)
            except Exception:
                pass
        ctx = ExecutionContext(
            guild_id=tenant_id,
            user_id=user_id,
            execution_id=execution.id,
            guild=guild,
            actor_member=actor,
            settings=settings,
        )
        res = await ActionDispatcher.dispatch(session, action_type, params, ctx)

        if res.requires_confirmation and res.confirmation_details:
            nonce = res.confirmation_details["action_nonce"]
            embed = DiscordUIComponents.create_confirmation_embed(
                res.confirmation_details["action_type"],
                res.confirmation_details["risk_level"],
                str(params)[:1500],
                nonce,
            )
            view = ConfirmationView(
                nonce=nonce,
                user_id=interaction.user.id,
                on_confirm=lambda itx: self._handle_confirm(itx, nonce),
            )
            await interaction.followup.send(embed=embed, view=view)
            await ConversationRepository.append(
                session,
                convo_id,
                "assistant",
                f"I asked for confirmation before doing {action_type}.",
            )
            return

        if res.success:
            detail = f"\n`{str(res.result)[:300]}`" if res.result else ""
            await interaction.followup.send(
                embed=DiscordUIComponents.create_success_embed(
                    "Action Executed", f"Successfully performed **{action_type}**.{detail}"
                )
            )
            await ConversationRepository.append(session, convo_id, "user", prompt)
            await ConversationRepository.append(
                session, convo_id, "assistant", f"I did {action_type} successfully."
            )
        else:
            await self._reply_error(
                interaction,
                session,
                convo_id,
                prompt,
                "Action Failed",
                res.error or "Execution failed.",
            )

    async def _run_plan(
        self, interaction, session, guild, tenant_id, user_id, prompt, convo_id, steps, settings
    ):
        execution = await ExecutionRepository.create_execution(
            session=session,
            guild_id=tenant_id,
            user_id=user_id,
            prompt=prompt,
            status="RUNNING",
            parsed_intent={"status": "action_plan", "steps": steps},
        )
        actor = interaction.user
        if not isinstance(actor, discord.Member):
            try:
                actor = await guild.fetch_member(interaction.user.id)
            except Exception:
                pass
        ctx = ExecutionContext(
            guild_id=tenant_id,
            user_id=user_id,
            execution_id=execution.id,
            guild=guild,
            actor_member=actor,
            settings=settings,
        )
        plan_res = await execute_action_plan(session, steps, 0, ctx)
        if not plan_res.get("completed") and plan_res.get("confirmation"):
            c = plan_res["confirmation"]
            embed = DiscordUIComponents.create_confirmation_embed(
                c["action_type"],
                c["risk_level"],
                f"Plan step {c['step_label']} needs approval.\n{summarize_plan(plan_res['results'])}",
                c["nonce"],
            )
            view = ConfirmationView(
                nonce=c["nonce"],
                user_id=interaction.user.id,
                on_confirm=lambda itx: self._handle_confirm(itx, c["nonce"]),
            )
            await interaction.followup.send(embed=embed, view=view)
            return
        summary = summarize_plan(plan_res["results"])
        all_ok = bool(plan_res["results"]) and all(r.get("ok") for r in plan_res["results"])
        await interaction.followup.send(
            embed=(
                DiscordUIComponents.create_success_embed("Plan Executed", summary)
                if all_ok
                else DiscordUIComponents.create_error_embed("Plan Had Failures", summary)
            )
        )
        await ConversationRepository.append(session, convo_id, "user", prompt)
        await ConversationRepository.append(session, convo_id, "assistant", summary[:500])

    # ---------- helpers ----------

    async def _reply_error(self, interaction, session, convo_id, prompt, title, msg):
        await interaction.followup.send(embed=DiscordUIComponents.create_error_embed(title, msg))
        try:
            await ConversationRepository.append(session, convo_id, "user", prompt)
            await ConversationRepository.append(
                session, convo_id, "assistant", f"That failed: {msg[:300]}"
            )
        except Exception:
            pass

    async def _ask_disambiguation(
        self,
        interaction,
        session,
        guild,
        tenant_id,
        user_id,
        prompt,
        convo_id,
        action_type,
        params,
        field,
        options,
        kind,
        default_channel_id,
        settings,
    ):
        pending = await PendingChoiceRepository.create(
            session,
            guild_id=tenant_id,
            user_id=user_id,
            action=action_type,
            params=params,
            field=field,
            options=options,
            prompt=prompt,
        )
        embed = DiscordUIComponents.create_disambiguation_embed(
            f"Which {kind}?",
            f"I found {len(options)} matching {kind}s. Pick one below (expires in 5 minutes).",
        )

        async def _picked(itx: discord.Interaction, chosen: str):
            async with AsyncSessionLocal() as s2:
                claimed = await PendingChoiceRepository.consume(s2, pending.id, tenant_id, user_id)
                if claimed is None:
                    await itx.response.send_message(
                        "Selection expired. Run /prompt again.", ephemeral=True
                    )
                    return
                valid_ids = {(o.get("id")) for o in (claimed.options or [])}
                if chosen not in valid_ids:
                    await itx.response.send_message("Invalid selection.", ephemeral=True)
                    return
                await itx.response.defer()
                await self._execute_single_action(
                    itx,
                    s2,
                    guild,
                    tenant_id,
                    user_id,
                    prompt,
                    convo_id,
                    claimed.action,
                    {**(claimed.params or {}), claimed.field: chosen},
                    settings,
                    default_channel_id,
                )

        view = DisambiguationView(options=options, user_id=interaction.user.id, on_pick=_picked)
        await interaction.followup.send(embed=embed, view=view)

    async def _resume_after_answer(
        self, interaction: discord.Interaction, pending_id: str, answer: str
    ):
        await interaction.response.defer(ephemeral=False)
        guild = interaction.guild
        if guild is None:
            await interaction.followup.send("Follow-ups only work within a server.", ephemeral=True)
            return
        guild_id = str(guild.id)
        user_id = str(interaction.user.id)
        async with AsyncSessionLocal() as session:
            await GuildRepository.find_or_create(session, guild_id, guild.name)
            pending = await PendingQuestionRepository.consume(
                session, pending_id, guild_id, user_id
            )
            if pending is None:
                await interaction.followup.send(
                    embed=DiscordUIComponents.create_error_embed(
                        "Answer Expired",
                        "This question already expired or belongs to someone else. Run /prompt again.",
                    )
                )
                return
            combined = f"{pending.originalPrompt}\nAdditional detail from user: {answer}"
            # Re-enter the flow as a fresh prompt turn.
            await self.handle_prompt(interaction, combined)

    async def _handle_confirm(self, interaction: discord.Interaction, nonce: str):
        await interaction.response.defer()
        guild = interaction.guild
        if guild is None:
            await interaction.followup.send(
                "Confirmations only work within a server.", ephemeral=True
            )
            return
        guild_id = str(guild.id)
        user_id = str(interaction.user.id)
        async with AsyncSessionLocal() as session:
            await GuildRepository.find_or_create(session, guild_id, guild.name)
            conf = await ConfirmationRepository.find_and_consume(
                session, nonce=nonce, guild_id=guild_id, user_id=user_id
            )
            if conf is None:
                await interaction.followup.send(
                    embed=DiscordUIComponents.create_error_embed(
                        "Invalid or Expired Confirmation",
                        "This confirmation expired, was consumed, or belongs to another user.",
                    )
                )
                return
            plan = conf.actionPlan or {}
            try:
                actor = interaction.user
                if not isinstance(actor, discord.Member):
                    actor = await guild.fetch_member(interaction.user.id)
                settings = await GuildRepository.get_settings(session, guild_id)
                ctx = ExecutionContext(
                    guild_id=guild_id,
                    user_id=user_id,
                    execution_id=f"conf-{nonce}",
                    guild=guild,
                    actor_member=actor,
                    settings=settings,
                )
                if plan.get("type") == PLAN_TYPE and isinstance(plan.get("steps"), list):
                    plan_res = await execute_action_plan(
                        session, plan["steps"], int(plan.get("index", 0)), ctx
                    )
                    if not plan_res.get("completed") and plan_res.get("confirmation"):
                        c = plan_res["confirmation"]
                        embed = DiscordUIComponents.create_confirmation_embed(
                            c["action_type"],
                            c["risk_level"],
                            f"Plan step {c['step_label']} needs approval.\n{summarize_plan(plan_res['results'])}",
                            c["nonce"],
                        )
                        view = ConfirmationView(
                            nonce=c["nonce"],
                            user_id=interaction.user.id,
                            on_confirm=lambda itx: self._handle_confirm(itx, c["nonce"]),
                        )
                        await interaction.followup.send(embed=embed, view=view)
                        return
                    summary = summarize_plan(plan_res["results"])
                    all_ok = bool(plan_res["results"]) and all(
                        r.get("ok") for r in plan_res["results"]
                    )
                    await interaction.followup.send(
                        embed=(
                            DiscordUIComponents.create_success_embed("Plan Executed", summary)
                            if all_ok
                            else DiscordUIComponents.create_error_embed(
                                "Plan Had Failures", summary
                            )
                        )
                    )
                    return
                action_def = ActionRegistry.get(plan.get("type", ""))
                if action_def is None:
                    await interaction.followup.send(
                        embed=DiscordUIComponents.create_error_embed(
                            "Execution Failed", "Action definition no longer exists."
                        )
                    )
                    return
                result = await action_def.handler(
                    ctx, action_def.input_schema(**(plan.get("input") or {}))
                )
                await interaction.followup.send(
                    embed=DiscordUIComponents.create_success_embed(
                        "Action Executed",
                        f"Successfully executed confirmed operation **{plan.get('type')}**.\n`{str(result)[:300]}`",
                    )
                )
            except Exception as err:
                await interaction.followup.send(
                    embed=DiscordUIComponents.create_error_embed("Execution Failed", str(err)[:500])
                )
